Ccsn temp

Inhyeok-Song opened the pull request:
I merged temperature/energy-based driver.
Now we can use temperature/energy-based driver as a compiling option.
For this, I slightly modify the Nuclear EoS Interface.

Inhyeok-Song committed:
Merge pull request #1 from gamer-project/master

Merge from gamer-project/master

Inhyeok-Song committed:
Merge remote-tracking branch 'upstream/master'

Inhyeok-Song committed:
Merge pull request #4 from hyschive/ccsn

Ccsn

Inhyeok-Song committed:
add findtemp/2 for temperature-based driver

Inhyeok-Song committed:
add nrho_mode, nye_mode and change NEPS->NTORE for switching temp/energy mode

Inhyeok-Song committed:
modifiy readtable for both temp/energy drvier

Inhyeok-Song committed:
modifiy the nuclear eos interface to use both temp/energy-based driver

Inhyeok-Song committed:
add temperature-based drvier

Inhyeok-Song committed:
modify some variables for temperature-based driver

Inhyeok-Song committed:
modify some variables for temperature-based driver

Inhyeok-Song committed:
deal 1.0e-10 and minor change in naming

Inhyeok-Song committed:
deal 1.0e-10 and minor change in naming

Inhyeok-Song committed:
deal with 1e-10 -> (int)floor()

Inhyeok-Song committed:
add compiling option temp/energy-based driver and temperature-based drvier source codes

Inhyeok-Song committed:
minor change

Inhyeok-Song committed:
minor change

Inhyeok-Song committed:
correct symbolic link

Inhyeok-Song committed:
fix duplicated define nuc_eos_C_linterp_some

Inhyeok-Song committed:
add NUC_TABLE_MODE_TEMP/ENGY compiling option macro

Inhyeok-Song committed:
delete useless

Inhyeok-Song committed:
minor change

Inhyeok-Song committed:
deal with 1e-10 -> floor()

hfhsieh commented:
I summarize my major comments here:
- Do not mix white spaces and tabs in indentation. 
- Check if `NUC_TABLE_MODE` is defined in case `EOS=EOS_NUCLEAR`.
- Check the coding style and alignment again.
- Run the Postbounce test problem with energy-based mode 
- For `NuclearEos.cpp`, I think it is possible to merge the code for different modes since most parts are similar. 
- I think it is possible to combine `findtemp.cpp` and `findenergy.cpp` into one function.
- The name `findtemp2.cpp` is not meaningful. 

I will check `findtemp2.cpp` in detail after the update.

hfhsieh commented on include/Macro.h:
```diff
@@ -79,6 +79,11 @@
+// nuclear eos table default mode
+#define NUC_TABLE_MODE_TEMP 0
+#define NUC_TABLE_MODE_ENGY 1
+
```
I think these two definitions are redundant since `NUC_TABLE_MODE_` is used as a compilation option.

Inhyeok-Song replied:
Then, where should I define these compiling options?

hfhsieh replied:
I think you do not need to define these two, but maybe I am wrong. I think it could be better to put these two at `NuclearEoS.h`.

Inhyeok-Song replied:
I add these here because other compiling options are here. Maybe Hsi-Yu gives better suggestion for the place.

hfhsieh commented on src/EoS/Nuclear/CPU_EoS_Nuclear.cpp:
```diff
@@ -21,28 +21,41 @@ __device__ static void EoS_General_Nuclear( const int Mode, real Out[], const re
+real   *g_logeps;
+real   *g_logtemp_mode;
+#endif
+
```
- Why remove the initialization of pointer to NULL?
- free `g_yes_mode` in `EoS_End_Nuclear()`
- alignment of `g_ntemp`, `*g_logtemp`, etc

Inhyeok-Song replied:
Added the NULL back and aligned the variables.

hfhsieh replied:
Also, initialize other pointes (e.g. `*g_logtemp`) to `NULL`.

Inhyeok-Song replied:
I initialized other pointers too.

hfhsieh commented on src/EoS/Nuclear/CPU_EoS_Nuclear.cpp:
```diff
@@ -217,8 +238,15 @@ static real EoS_DensEint2Pres_Nuclear( const real Dens_Code, const real Eint_Cod
    real Temp_MeV  = NULL_REAL;
    real Entr      = NULL_REAL;
    real Pres_CGS  = NULL_REAL;
+   real mu_nu     = NULL_REAL;
```
`mu_nu` is not dimensionless. Maybe it is better to specify the unit in the name.

Inhyeok-Song replied:
I removed `mu_nu` for the time being since it needed for source terms. I will update this when I add the source terms.

hfhsieh commented on src/EoS/Nuclear/CPU_EoS_Nuclear.cpp:
```diff
@@ -217,8 +238,15 @@ static real EoS_DensEint2Pres_Nuclear( const real Dens_Code, const real Eint_Cod
+#else
+   const real Tolerance = 1e-6;
+#endif
 
```
Would it be better to define `Tolerance` in `NuclearEoS.h`?

Inhyeok-Song replied:
I agree with you. I'd better to do that.

hfhsieh replied:
For the Tolerance moved to `NuclearEoS.h`:
1. Typo in the comment: "Torlerance" -> "Tolerance"
2. The leading three spaces seem to be not necessary.

Inhyeok-Song replied:
I corrected the typo and removed unnecessary space in `NuclearEoS.h`

hfhsieh commented on src/EoS/Nuclear/CPU_EoS_Nuclear.cpp:
```diff
@@ -227,9 +255,9 @@ static real EoS_DensEint2Pres_Nuclear( const real Dens_Code, const real Eint_Cod
+   //if ( Nuc_Overflow(sEint_CGS) )
+   //   printf( "ERROR : EoS overflow (sEint_CGS %13.7e, Eint_Code %13.7e, Dens_Code %13.7e, sEint2CGS %13.7e) in %s() !!\n",
+   //           sEint_CGS, Eint_Code, Dens_Code, sEint2CGS, __FUNCTION__ );
 
```
Why comment out the check of `sEint_CGS`?

Inhyeok-Song replied:
I added back the check of `sEint_CGS`

hfhsieh commented on src/EoS/Nuclear/CPU_EoS_Nuclear.cpp:
```diff
@@ -238,18 +266,19 @@ static real EoS_DensEint2Pres_Nuclear( const real Dens_Code, const real Eint_Cod
+                    EnergyShift, NRho, NTorE, NYe, NRho_Mode, NMode, NYe_Mode,
+                    Table[NUC_TAB_ALL], Table[NUC_TAB_ALL_MODE], Table[NUC_TAB_RHO], Table[NUC_TAB_TORE], Table[NUC_TAB_YE], 
+                    Table[NUC_TAB_RHO_MODE], Table[NUC_TAB_ENTE_MODE], Table[NUC_TAB_ENTR_MODE],  Table[NUC_TAB_PRES_MODE],
+                    Table[NUC_TAB_YE_MODE], Mode, &Err, Tolerance );
```
Remove one redundant space before `Table[NUC_TABL_PRES_MODE]`, and other places.

Inhyeok-Song replied:
I removed it.

hfhsieh commented on src/EoS/Nuclear/CUAPI_PassNuclearEoSTable2GPU.cu:
```diff
@@ -6,9 +6,16 @@
+extern int    g_ntemp;
+#elif ( NUC_TABLE_MODE == NUC_TABLE_MODE_ENGY )
+extern int    g_neps;
+#endif
```
alignment

Inhyeok-Song replied:
alignment is done.

hfhsieh commented on src/EoS/Nuclear/EoS_End_Nuclear.cpp:
```diff
@@ -11,11 +11,19 @@ void CUAPI_MemFree_NuclearEoS();
+#elif ( NUC_TABLE_MODE == NUC_TABLE_MODE_ENGY )
+extern real   *g_logeps;
+extern real   *g_logtemp_mode;
+#endif
```
alignment

hfhsieh replied:
Maybe you can remove one more space after `real` to make these four variables also align with the others above.

Inhyeok-Song replied:
I removed the useless space.

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.h:
```diff
@@ -21,20 +21,24 @@
-#define NUC_TAB_ENTR_MODE     6     // entr_mode
-#define NUC_TAB_PRES_MODE     7     // logprss_mode
+#define NUC_TAB_RHO_MODE      5     // logrho_mode 
+#define NUC_TAB_ENTE_MODE     6     // logenergy_mode/logtemp_mode
```
Would it be better to use `NUC_TAB_TORE_MODE` instead of `NUC_TAB_ENTE_MODE`

Inhyeok-Song replied:
Yes. Maybe `NUC_TAB_EORT_MODE` is better since T-based drvier uses `NUC_TAB_ENGY_MODE` and e-based drive uses `NUC_TAB_TEMP_MODE`.

hfhsieh replied:
I suggest "TORE", instead of "EORT" (and "EN TE"), because other related variables use "TorE" (e.g., `NUC_TAB_TORE`)

hfhsieh commented on src/EoS/Nuclear/cubinterp_some.cpp:
```diff
@@ -5,8 +5,11 @@
+void nuc_eos_C_linterp_some( const real x, const real y, const real z,
+                             real *output_vars, const real *alltables,
+                             const int nx, const int ny, const int nz, const int nvars,
+                             const real *xt, const real *yt, const real *zt );
```
Just curious, why replace `include` by prototype?

Inhyeok-Song replied:
Since this causes the multiple definition issue when I compiled.

hfhsieh commented on src/EoS/Nuclear/cubinterp_some.cpp:
```diff
@@ -25,6 +28,7 @@ void nuc_eos_C_linterp_some( const real x, const real y, const real z,
 
 
 
+
```
Remove this redundant line

hfhsieh commented on src/EoS/Nuclear/cubinterp_some.cpp:
```diff
@@ -90,9 +94,9 @@ void nuc_eos_C_cubinterp_some( const real x, const real y, const real z,
-   iz = (int)( ( z - zt[0] + (real)1.0e-10 )*dzi );
+   ix = (int)floor( ( x - xt[0] )*dxi );
+   iy = (int)floor( ( y - yt[0] )*dyi );
+   iz = (int)floor( ( z - zt[0] )*dzi );
```
- Here and other places.
  - In fact, `floor()` is not necessary as Schive corrected on slack
  - I am wondering whether removing `1.0e-10` could result in any inconsistency since it is retained in the code for table construction. Maybe you can check whether the table generated with/without `1e-10` is the same.
- Should we handle the case `i[xyz] = n[xyz] - 1`, or do you think it is not necessary? I ask this the [item](https://github.com/kuochuanpan/gamer-supernovae/issues/1#issue-924540581) is checked but no change is found here.

Inhyeok-Song replied:
Actually I found no difference between previous version with `1e-10` and this version with `floor` in any cases (driver and table constructions). But I changed this because I think this looks more reasonable and straightforward.

hfhsieh replied:
- I suggest removing `floor()` because the effect of `(int) ...` should be the same as `(int) floor(...)`, but faster because one operation less.
- Any comment for the question on `i[xyz]`?

Inhyeok-Song replied:
- I removed `floor()`.
- `i[xyz]` this problem would (rarely/never) happen when we add `1e-10`. Since now we removed `1e-10` this won't be happened.

hfhsieh commented on src/EoS/Nuclear/readtable.cpp:
```diff
@@ -24,19 +24,30 @@
+extern int    g_neps;
+extern real   *g_logeps;
+extern real   *g_logtemp_mode;
+#endif
```
alignment

hfhsieh commented on src/EoS/Nuclear/readtable.cpp:
```diff
@@ -109,60 +120,95 @@ void nuc_eos_C_ReadTable( char *nuceos_table_name )
+      Aux_Error( ERROR_INFO, "cannot allocate memory for EOS table !!\n");
 
-   if (  ! ( alltables_mode_temp = (real*)malloc(g_nrho*g_nmode*g_nye*3            *sizeof(real)) )  )
+   if (  ! ( alltables_mode_temp = (real*)malloc(g_nrho_mode*g_nmode*g_nye_mode*3   *sizeof(real)) )  )
```
alignment of `sizeof` and `)`, and other places

hfhsieh commented on src/EoS/Nuclear/readtable.cpp:
```diff
@@ -187,18 +233,31 @@ void nuc_eos_C_ReadTable( char *nuceos_table_name )
+#if   ( NUC_TABLE_MODE == NUC_TABLE_MODE_TEMP )
+   READ_EOSTABLE_MODE_HDF5( "logtemp_ener", 0 );
+   READ_EOSTABLE_MODE_HDF5( "logtemp_entr", 1 );
+   READ_EOSTABLE_MODE_HDF5( "logtemp_prss", 2 );
```
alignment

hfhsieh commented on src/EoS/Nuclear/readtable.cpp:
```diff
@@ -207,23 +266,23 @@ void nuc_eos_C_ReadTable( char *nuceos_table_name )
-   for (int i=0; i<g_nrho; i++)
+   for (int k=0; k<g_nye;   k++)
+   for (int j=0; j<n_def_mode; j++)
+   for (int i=0; i<g_nrho;  i++)
```
alignment of `[ijk]++`

hfhsieh commented on src/EoS/Nuclear/readtable.cpp:
```diff
@@ -207,23 +266,23 @@ void nuc_eos_C_ReadTable( char *nuceos_table_name )
-   for (int i=0; i<g_nrho;  i++)
+   for (int k=0; k<g_nye_mode;   k++)
+   for (int j=0; j<g_nmode;      j++)
+   for (int i=0; i<g_nrho_mode;  i++)
```
Remove the redundant space after ;

hfhsieh commented on src/EoS/Nuclear/readtable.cpp:
```diff
@@ -109,60 +120,95 @@ void nuc_eos_C_ReadTable( char *nuceos_table_name )
 
 // allocate memory for tables
    real *alltables_temp      = NULL;
    real *alltables_mode_temp = NULL;
```
Would it be better to rename these two `alltables` variabeles since it is no longer for temperature only?

Inhyeok-Song replied:
sorry for the confusement this alltabless_temp means temporary array for alltables. Later we change the indices and store in alltables

hfhsieh replied:
How about using `_tmp`?

Inhyeok-Song replied:
sounds great.

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -23,122 +40,213 @@ void findenergy( const real x, const real y, const real z,
 
-
-
-
```
add three empty lines here

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -23,122 +40,213 @@ void findenergy( const real x, const real y, const real z,
+//                                  112 : entropy too low  (energy-based table)
+//                                  113 : log(P) too high  (energy-based table)
+//                                  114 : log(P) too low   (energy-based table)
+//                rfeps           : tolerence for interpolations
```
- Why modify the structure, e.g. the (1) and (2) in line 34?
- Why change the leading characters from upper case to lower case?
- error code 107 - 114 would only occur in the energy-based table?
- some error code is missing here, e.g., 2xx, 66x

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -23,122 +40,213 @@ void findenergy( const real x, const real y, const real z,
 
    if ( xye > yes  [nye -1] )  {  *keyerr = 101;  return;  }
    if ( xye < yes  [     0] )  {  *keyerr = 102;  return;  }
-   if ( xye != xye )           {  *keyerr = 201;  return;  }
```
Why remove the sanity check of NaN for density and ye, and checks in other places?

Inhyeok-Song replied:
I added back.

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -163,12 +270,11 @@ void nuc_eos_C_short( const real xrho, real *xenr, const real xye,
+         if ( lprs > logprss_mode[nmode-1] ) {  *keyerr = 113;  return;  }
+         if ( lprs < logprss_mode[      0] ) {  *keyerr = 114;  return;  }
 
          findenergy( lr, lprs, xye, &leps, alltables_mode, nrho, nmode, nye, neps,
```
I think we should use ``nrho_mode` and `nye_mode` here, right?

Inhyeok-Song replied:
you are right.

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -23,122 +40,213 @@ void findenergy( const real x, const real y, const real z,
+
+   const int  neps    = ntoreps;
+   const real *logeps = logtoreps;
+   real       leps    = NULL_REAL;
```
alignment

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -195,11 +301,12 @@ void nuc_eos_C_short( const real xrho, real *xenr, const real xye,
    *xmunu = res[3];
    *xcs2  = res[4];
 
+#endif // #elif Table_Mode == TABLE_MODE_ENGY
```
Maybe `#if NUC_TABLE_MODE ... else ...` is better?

Inhyeok-Song replied:
Yes. That would be better.

hfhsieh commented on src/EoS/Nuclear/findtemp.cpp:
```diff
@@ -0,0 +1,337 @@
+
+//-------------------------------------------------------------------------------------
+// Function    :  findtemp
+// Description :  Finding energy from different modes
```
Finding "temperature"

hfhsieh commented on src/EoS/Nuclear/findtemp.cpp:
```diff
@@ -0,0 +1,337 @@
+//                                 --> 1: Energy mode   (coming in with internal energy)
+//                                     2: Entropy mode  (coming in with entropy)
+//                                     3: Pressure mode (coming in with P)
+//                keyerr         : output error
```
- Format is not consistent with `findenergy.cpp`
  - e.g., upper/lower case of leading characters
- missing description for return parameters

- same questions for `findtemp_brdy()`

Inhyeok-Song replied:
I corrected all. please check.

hfhsieh replied:
You can use the command "diff" to find out the difference:

1. duplicate description "Using 3D Catmull-Rom ..."
2. Following (1), Use linear ... search the corresponding "temperature". Check `findtemp_bdry()` too.
3. "output error" -> "Output error"
4. alignment of "Return ..."
5. The symbol of entropy use `e` in the Note, but `entropy` for the argument `keymode`. CHeck other places too.
6. Correct "Finding energy ..." to "Finding temperature ..." in `findtemp_bdry()`

Inhyeok-Song replied:
I checked above comments.

Inhyeok-Song committed:
correct format and comment

Inhyeok-Song committed:
Update CPU_EoS_Nuclear.cpp

I added back NULL and aligned the variables.

Inhyeok-Song committed:
Update readtable.cpp

Inhyeok-Song committed:
Update CPU_EoS_Nuclear.cpp

Inhyeok-Song committed:
Update NuclearEoS.h

added torlerance for Newton-Raphson or bisection method in temperature driver

Inhyeok-Song committed:
Update CPU_EoS_Nuclear.cpp

added back check Nuc_Overflow(sEint_CGS)

Inhyeok-Song committed:
Update CPU_EoS_Nuclear.cpp

remove redundant spaces.

Inhyeok-Song committed:
Update CUAPI_PassNuclearEoSTable2GPU.cu

aligned the variables

Inhyeok-Song committed:
Update EoS_End_Nuclear.cpp

add free( g_yes_mode) and aligned the variables.

Inhyeok-Song committed:
Update findtemp.cpp

corrected comment.

Inhyeok-Song committed:
Update readtable.cpp

rename the variables `alltables_temp`->`alltables_tmp`, `alltables_mode_temp`->`alltables_mode_tmp`

Inhyeok-Song committed:
Update NuclearEoS.cpp

Inhyeok-Song committed:
Update cubinterp_some.cpp

Inhyeok-Song committed:
Update readtable.cpp

change the macro name.

Inhyeok-Song committed:
Update CPU_EoS_Nuclear.cpp

change the macro name

Inhyeok-Song committed:
Update CUAPI_PassNuclearEoSTable2GPU.cu

change the macro name

Inhyeok-Song committed:
Update NuclearEoS.h

change the macro name

hfhsieh commented on src/EoS/Nuclear/CPU_EoS_Nuclear.cpp:
```diff
@@ -110,9 +123,15 @@ void EoS_SetAuxArray_Nuclear( double AuxArray_Flt[], int AuxArray_Int[] )
+   AuxArray_Int[NUC_AUX_NTORE     ] = g_ntemp;
+#elif ( NUC_TABLE_MODE == NUC_TABLE_MODE_ENGY )
+   AuxArray_Int[NUC_AUX_NTORE     ] = g_neps;
+#endif
```
It is better to align `#if` with other lines (see line 203 for example). Check other places too.

Inhyeok-Song replied:
aligned.

hfhsieh commented on src/EoS/Nuclear/findtemp.cpp:
```diff
@@ -0,0 +1,343 @@
+   int iv;
+   if      ( keymode == NUC_MODE_ENGY ) iv = 0; // energy table for the energy mode
+   else if ( keymode == NUC_MODE_ENTR ) iv = 1; // energy table for the entropy mode
+   else if ( keymode == NUC_MODE_PRES ) iv = 2; // energy table for the pressure mode
```
1. I think it should be "temperature" table for ...?  Check other places too.
2. The number spacing before `iv`. You use 1 space here and 2 spaces in `findtemp_bdry()`, but 3 spaces are used in `findenergy.cpp`

Inhyeok-Song replied:
corrected

hfhsieh commented on src/EoS/Nuclear/readtable.cpp:
```diff
@@ -109,60 +120,95 @@ void nuc_eos_C_ReadTable( char *nuceos_table_name )
 
-   if (  ! ( alltables_mode_temp = (real*)malloc(g_nrho*g_nmode*g_nye*3            *sizeof(real)) )  )
+   if (  ! ( alltables_mode_tmp  = (real*)malloc(g_nrho_mode*g_nmode*g_nye_mode*3      *sizeof(real)) )  )
       Aux_Error( ERROR_INFO, "cannot allocate memory for EOS table !!\n" );
```
For some `Aux_Error()` here, one space is missed before `)`. Or you can replace them with this one.

Inhyeok-Song replied:
I corrected.

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -7,14 +7,31 @@
 #include "findenergy.cu"
+#include "findtemp.cu"
+#include "findtemp2.cu"
+
```
Remove this redundant empty line.

Inhyeok-Song replied:
removed.

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -30,115 +47,217 @@ void findenergy( const real x, const real y, const real z,
+//                                  201 : lr  has NaN value
+//                                  203 : xye has NaN value
+//                rfeps           : Tolerence for interpolations
 //-----------------------------------------------------------------------------------------------
```
Check the function documentation again:
1. remove the trailing space at line 50
2. some error codes are still missed: 20x series and 65x series
3. Erro with code 111 - 114 will only happen in the energy-based table, never in the temperature-based table?

To me, this documentation is somehow specific to the temperature-based table because
1. On Note (3), is the temperature mode the default mode in the energy-based table?
2. For `nmode`,  no size of log(T) for energy-based table
3. etc...

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -195,11 +312,12 @@ void nuc_eos_C_short( const real xrho, real *xenr, const real xye,
 } // FUNCTION : nuc_eos_C_short
 
 
-
```
Add this empty line back.

Inhyeok-Song replied:
added the empty line back.

hfhsieh commented on src/EoS/Nuclear/NuclearEoS.cpp:
```diff
@@ -195,11 +312,12 @@ void nuc_eos_C_short( const real xrho, real *xenr, const real xye,
    *xmunu = res[3];
    *xcs2  = res[4];
 
+#endif // #elif NUC_TABLE_MODE ... else ...
```
`#if NUC_TABLE_MODE ...`

Inhyeok-Song replied:
maybe outdated.

hfhsieh reviewed (commented):
Here I add more comments / suggestions. Besides the suggestions not responded to in the last review, 

1. It could be better to check `NUC_TABLE_MODE` is defined and its value is `NUC_TABLE_MODE_ENGY` or `NUC_TABLE_MODE_TEMP` when nuclear EoS is adopted. See `Aux_Check_Parameter.cpp` for example.
2. I suggest recording additional information in `Record_Note`, e.g.
  - temperature- / energy-based table
  - interpolation scheme

I did not check `findtemp2.cpp` because there are lots of inconsistencies in the coding style. Besides that, 
1. Could it be better to use `[xyz]` instead of `[rho / temp / ye]` like that in `findtemp.cpp`
2. Is `DEBUG` still needed?

Also, I think there are too many duplicates of interpolation (despite there are slightly different). For example, the cubic scheme can be found in `findtemp.cpp`, `findenergy.cpp`, `cubinterp.cpp`, and linear scheme in `findtemp.cpp`, `findtemp2.cpp`, `findenergy.cpp`, `linterp_some.cpp`.

Inhyeok-Song committed:
Update CPU_EoS_Nuclear.cpp

checked the indentation and initializing pointers.

Inhyeok-Song committed:
Update EoS_End_Nuclear.cpp

minor update

Inhyeok-Song committed:
Update cubinterp_some.cpp

remove floor().

Inhyeok-Song committed:
Update findenergy.cpp

remove `floor()`

Inhyeok-Song committed:
Update NuclearEoS.cpp

minor update.

Inhyeok-Song committed:
Update NuclearEoS.h

minor update.

Inhyeok-Song committed:
Update findtemp.cpp

modified the descriptions and removed `floor()`

Inhyeok-Song committed:
Update findtemp.cpp

minor update.

Inhyeok-Song committed:
Update readtable.cpp

minor update.

Inhyeok-Song committed:
Update cubinterp_some.cpp

missing `(`

Inhyeok-Song committed:
Update linterp_some.cpp

remove `floor()`

Inhyeok-Song committed:
changed the coding style

changed the coding style.

hyschive commented:
Test slack integration

hfhsieh commented:
One more potential bug here. For MHD runs, the program will crash during dumping the HDF5 output, in both TEMP and ENGY modes. The adopted EoS tables are generated from your `EoS Converter` with version `df09154`. The error message is 

~~~
Output_DumpData_Total_HDF5 (DumpID = 0) ...
HDF5-DIAG: Error detected in HDF5 (1.8.21) MPI-process 0:
  #000: H5Shyper.c line 6614 in H5Sselect_hyperslab(): not a data space
    major: Invalid arguments to routine
    minor: Inappropriate type
********************************************************************************
ERROR : failed to create a hyperslab for the magnetic field !!
        Rank <0>, file <Output/Output_DumpData_Total_HDF5.cpp>, line <1371>, function <Output_DumpData_Total_HDF5>
********************************************************************************
~~~

However, using the `hyschive:ccsn` branch with the table generated on May, the program works well during dumping the HDF5 output. Thus, there is something wrong in either this PR or your code for table converter, or elsewhere.

Inhyeok-Song committed:
delete outdated files

Inhyeok-Song committed:
merge findtemp, findenergy and eos mode

Inhyeok-Song committed:
change source codes for nuclear EoS

Inhyeok-Song committed:
minor update for energy-based drvier

Inhyeok-Song committed:
add a runtime parameter linear/cubic interpolation schemes

Inhyeok-Song committed:
minor update

Inhyeok-Song committed:
remove mu_nu for the time being

Inhyeok-Song committed:
minor update

Inhyeok-Song committed:
minor update

Inhyeok-Song committed:
deal with min, max(ix, iy, iz)

Inhyeok-Song committed:
add nieve initial guess of T and minor update

Inhyeok-Song committed:
minor fix

Inhyeok-Song committed:
adding check for nuclear eos table types

Inhyeok-Song committed:
add NUC_TABLE_MODE

Inhyeok-Song committed:
add NUC_TABLE_MODE

Inhyeok-Song committed:
removed wrong commenting out

hfhsieh commented:
## To update:
1. control the interpolation schemes used in `findtoreps()` and `nuc_eos_C_xx_some()` separately
2. update the `FormatVersion` in HDF5 output (line 1730 in `Output_DumpData_Total_HDF5.cpp`)
3. use more meaningful output in `Aux_TakeNote()`:
  -  for `NUC_TABLE_MODE`, see part (a) for example since it is a compiler option
  - for `NUC_EOS_INTERPOL_SCHEME`, see `OPT__LR_LIMITER` for example
4. In `findtemp_NR_bisection.cpp`, replace `fabs` with `FABS` to support both float/double precisions

## To be discussed:
### on these `#     if DEBUG` in `findtemp_NR_bisection.cpp`

Do we need these debug information? If yes, use `GAMER_DEBUG` instead, and output the information formally. Or I will remove all of them in my revised version.

### on `i[xyz]` (see also https://github.com/hyschive/gamer-fork/pull/20#discussion_r673075178)

I still do not figure out why not replace `ix == nx-2` with `ix >= nx-2` (and so for other) to handle the potential issue. Also, removing the subtraction of 1e-10 would make `ix = nx-1` more possibly occurs. 

In the current PR, you add `ix = MAX( 1, MIN( ix, nx-2 ) );` to restrict the value of `ix`, but is the constraint reasonable? For example, after this statement the possible minimum value of `ix` is 1, and hence should we still need to consider the condition `ix == 0` in the following code for linear interpolation at boundaries?

## Things I will revise in next PR
- coding style
- function description...
- inconsistent code with current `ccsn` branch
- optimization of nuclear EoS driver

Inhyeok-Song committed:
seperate interpolation schemes and fixed potential issue

Inhyeok-Song committed:
remove stroing interpolation schemes in HDF5 format temporarily

Inhyeok-Song commented:
> ## To update:
> 1. control the interpolation schemes used in `findtoreps()` and `nuc_eos_C_xx_some()` separately
> 2. update the `FormatVersion` in HDF5 output (line 1730 in `Output_DumpData_Total_HDF5.cpp`)
> 3. use more meaningful output in `Aux_TakeNote()`:
> 
> * for `NUC_TABLE_MODE`, see part (a) for example since it is a compiler option
> * for `NUC_EOS_INTERPOL_SCHEME`, see `OPT__LR_LIMITER` for example
> 
> 1. In `findtemp_NR_bisection.cpp`, replace `fabs` with `FABS` to support both float/double precisions
> 
> ## To be discussed:
> ### on these `# if DEBUG` in `findtemp_NR_bisection.cpp`
> Do we need these debug information? If yes, use `GAMER_DEBUG` instead, and output the information formally. Or I will remove all of them in my revised version.
> 
> ### on `i[xyz]` (see also [#20 (comment)](https://github.com/hyschive/gamer-fork/pull/20#discussion_r673075178))
> I still do not figure out why not replace `ix == nx-2` with `ix >= nx-2` (and so for other) to handle the potential issue. Also, removing the subtraction of 1e-10 would make `ix = nx-1` more possibly occurs.
> 
> In the current PR, you add `ix = MAX( 1, MIN( ix, nx-2 ) );` to restrict the value of `ix`, but is the constraint reasonable? For example, after this statement the possible minimum value of `ix` is 1, and hence should we still need to consider the condition `ix == 0` in the following code for linear interpolation at boundaries?
> 
> ## Things I will revise in next PR
> * coding style
> * function description...
> * inconsistent code with current `ccsn` branch
> * optimization of nuclear EoS driver

>## About the things to be discussed
* I removed DEBUG parts in findtemp_NR_bisection.cpp

* This is for checking whether this point is at the very next to the boundaries before the cubic interpolation . If the point is at a boundary I will go for linear interpolation. Furthermore, before findtoreps we check the whether the point is in the parameter space or not in NuclearEoS.cpp . So I think there will be no difference between ix>=nx-1 and ix==nx. I just used the expression which I think more straighforward. Anyway I changed this part also. And the subtraction of 1e-10 is no longer needed. As I said we check the point before this function and also restrict the value at findtoreps_bdry.
* For ix = MAX( 1, MIN( ix, nx-2 ) ); , ix ranges from 1 to nx-1. ix=0 would mean it is lower than the lower boundary. Sorry for the confusion. In `findtoreps`, `ix` ranges from `0` to `nx-2`. In `findtoreps_bdry`, `ix` ranges from `1` to `nx-1`.

Inhyeok-Song committed:
change the names

Inhyeok-Song committed:
add NaN checks

Inhyeok-Song committed:
minor change

Inhyeok-Song committed:
minor change

Inhyeok-Song committed:
check indices

Inhyeok-Song committed:
restrict indices

Inhyeok-Song committed:
wrong indices

Inhyeok-Song committed:
fixing index issues

Inhyeok-Song committed:
add interpolation scheme macros for nuclear EOS

Inhyeok-Song committed:
fixed bugs in eos interpolations schemes

Inhyeok-Song committed:
move interpolation scheme macros to Macro.h

Inhyeok-Song committed:
update for taking note of Nuclear EoS interpolation schemes

hyschive commented on include/Macro.h:
```diff
@@ -79,6 +79,11 @@
 
+// nuclear eos table default mode
+#define NUC_TABLE_MODE_TEMP 0
+#define NUC_TABLE_MODE_ENGY 1
```
Change to the following since we typically use `0` for "off".
```
#define NUC_TABLE_MODE_TEMP 1
#define NUC_TABLE_MODE_ENGY 2
```

hyschive commented on include/Typedef.h:
```diff
@@ -402,6 +402,15 @@ const PatchType_t
+typedef int NucEoSInterpol_t;
+const NucEoSInterpol_t
+   NUC_INTERPOL_LINEAR = 0,
+   NUC_INTERPOL_CUBIC  = 1;
```
Use `1, 2` instead of `0, 1`

hyschive commented on src/Init/Init_Load_Parameter.cpp:
```diff
@@ -229,7 +229,9 @@ void Init_Load_Parameter()
 #  if ( EOS == EOS_NUCLEAR )
-   ReadPara->Add( "NUC_TABLE",                  NUC_TABLE,                        Useless_str,     Useless_str,   Useless_str    );
+   ReadPara->Add( "NUC_TABLE",                   NUC_TABLE,                        Useless_str,     Useless_str,   Useless_str    );
+   ReadPara->Add( "NUC_EOS_INTERPOL_TL",        &NUC_EOS_INTERPOL_TL,              0,               0,             1              );
+   ReadPara->Add( "NUC_EOS_INTERPOL_OTHER",     &NUC_EOS_INTERPOL_OTHER,           0,               0,             1              );
```
Replace `(0, 1)` by `(1, 2)`

hyschive reviewed (changes requested):
@Inhyeok-Song Good job. I only leave some minor comments.

Inhyeok-Song committed:
change options (0, 1)->(1, 2)
